### PR TITLE
Add user access history

### PR DESCRIPTION
**Pull request by g-theriot**

### CODE REVIEW

**Comment by dastan-rahimi** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 26
- Added 34
           

Clones removed
==============
+ LEAF_Nexus/sources/Group.php  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/VALEAF/LEAF/pullRequest?prid=4809995&bid=15999835)

**Comment by dastan-rahimi** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 26
- Added 33
           

Clones removed
==============
+ LEAF_Nexus/sources/Group.php  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/VALEAF/LEAF/pullRequest?prid=4809995&bid=15999835)

File: LEAF_Nexus/db_upgrade/Update_OC_DB_5204-2019110100.sql [hunk 1]
```diff
@@ -6,6 +6,7 @@ CREATE TABLE `data_action_log` (
   `action` varchar(45) DEFAULT NULL,
   `userID` int(11) DEFAULT NULL,
   `timestamp` DATETIME DEFAULT NULL,
+  `userDisplay` VARCHAR(255) NUL
```
Comment (thomasnguyen95) [inline LEAF_Nexus/db_upgrade/Update_OC_DB_5204-2019110100.sql]: ```NULL```
not ```NUL```

**Comment by dastan-rahimi** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 29
- Added 44
           

Clones removed
==============
+ LEAF_Nexus/sources/Group.php  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/VALEAF/LEAF/pullRequest?prid=4809995&bid=15999835)

File: .vs/ProjectSettings.json [hunk 2]
```diff
@@ -0,0 +1,3 @@
+{
+  "CurrentProjectSetting": null
```
Comment (dastan-rahimi) [inline .vs/ProjectSettings.json]: Remove unrelated files

File: .vs/VSWorkspaceState.json [hunk 3]
```diff
@@ -0,0 +1,10 @@
+{
+  "ExpandedNodes": [
+    "",
+    "\\libs",
+    "\\libs\\logger",
```
Comment (dastan-rahimi) [inline .vs/VSWorkspaceState.json]: Remove

File: .vs/ProjectSettings.json (hunk 2 above)
Comment (thomasnguyen95) [reply .vs/ProjectSettings.json]: or added .vs/ to .gitignore so this content doesn't get staged.

File: LEAF_Request_Portal/admin/Group.php [hunk 4]
```diff
@@ -94,6 +102,11 @@ public function addMember($member, $groupIDs)
                               ':groupID' => (int)$group, );
                 $res = $this->db->prepared_query('INSERT INTO users (userID, groupID)
                                                     VALUES (:userID, :groupID)', $vars);
+                
+                $this->dataActionLogger->logAction(\DataActions::ADD,\LoggableTypes::EMPLOYEE,[
+                    new LogItem("users","userID", $member, $this->getEmployeeDisplay($member)),
```
Comment (thomasnguyen95) [inline LEAF_Request_Portal/admin/Group.php]: nit-pick: Should there be inconsistencies with the use of ```/LogItem``` usage here vs in ```LEAF_Nexus/sources/Group.php```

File: LEAF_Request_Portal/admin/Group.php [hunk 5]
```diff
@@ -134,4 +152,31 @@ public function getGroupsAndMembers()
 
         return $list;
     }
+
+    public function getGroupName($groupId){
+        $vars = array(":groupID"=> $groupId);
+        $res = $this->db->prepared_query('SELECT * FROM `groups` WHERE groupID = :groupID', $vars);
```
Comment (thomasnguyen95) [inline LEAF_Request_Portal/admin/Group.php]: We're only looking for name here, we don't need to do a ```Select *```

File: LEAF_Request_Portal/admin/templates/mod_form.tpl [hunk 6]
```diff
@@ -160,20 +193,29 @@ function editProperties(isSubForm) {
...
+
+            if(visibleChanged){
                 $.ajax({
                     type: 'POST',
                     url: '../api/?a=formEditor/formVisible',
                     data: {visible: $('#visible').val(),
                         categoryID: currCategoryID,
                         CSRFToken: '<!--{$CSRFToken}-->'},
                     success: function(res) {
+                        categories[currCategoryID].visiblev= $('#visible').val();
```
Comment (thomasnguyen95) [inline LEAF_Request_Portal/admin/templates/mod_form.tpl]: Should this be ```visiblev```? I don't think it's being used against after being defined.

File: LEAF_Request_Portal/admin/templates/mod_form.tpl [hunk 7]
```diff
@@ -1823,6 +1933,24 @@ function renderSecureFormsInfo(res) {
     }
 }
 
+function viewHistory(categoryId){
+    dialog_simple.setContent('');
+    dialog_simple.setTitle('Form History');
+	dialog_simple.show();
```
Comment (thomasnguyen95) [inline LEAF_Request_Portal/admin/templates/mod_form.tpl]: np: spacing

File: LEAF_Request_Portal/admin/templates/mod_groups.tpl [hunk 8]
```diff
@@ -203,6 +205,24 @@ function getGroupList() {
     });
 }
 
+function viewHistory(groupID){
+    dialog_simple.setContent('');
+    dialog_simple.setTitle('Group History');
+	dialog_simple.show();
```
Comment (thomasnguyen95) [inline LEAF_Request_Portal/admin/templates/mod_groups.tpl]: np: spacing

File: LEAF_Request_Portal/admin/templates/mod_svcChief.tpl [hunk 9]
```diff
@@ -136,7 +136,9 @@ function initiateWidget(serviceID) {
                     if(serviceID > 0) {
                         button_deleteGroup = '';
                     }
-                    dialog.setContent('<div id="employees"></div><br /><h3>Add Employee:</h3><div id="employeeSelector"></div><br /><br />' + button_deleteGroup);
+                    dialog.setContent(
+                        '<button style="float:right" class="buttonNorm" onclick="viewHistory('+serviceID+')"><img src="../../libs/dynicons/?img=appointment.svg&amp;w=16" alt="View Status" title="View History" style="vertical-align: middle"> View History</button>'+
```
Comment (thomasnguyen95) [inline LEAF_Request_Portal/admin/templates/mod_svcChief.tpl]: Please change Alt Text to be ```View History```

File: LEAF_Request_Portal/admin/templates/mod_svcChief.tpl [hunk 10]
```diff
@@ -239,6 +241,24 @@ function getGroupList() {
 	});
 }
 
+function viewHistory(groupID){
+    dialog_simple.setContent('');
+    dialog_simple.setTitle('Service Chief History');
+	dialog_simple.show();
```
Comment (thomasnguyen95) [inline LEAF_Request_Portal/admin/templates/mod_svcChief.tpl]: np: spacing

File: LEAF_Request_Portal/sources/FormEditor.php [hunk 11]
```diff
@@ -316,39 +466,78 @@ public function setFormNeedToKnow($categoryID, $input)
...
+
+        return $result;
     }
 
     public function setFormType($categoryID, $input){
 
-      $vars = array(':categoryID' => $categoryID,
-          ':input' => $input, );
+        $vars = array(':categoryID' => $categoryID,
+            ':input' => $input, );
```
Comment (thomasnguyen95) [inline LEAF_Request_Portal/sources/FormEditor.php]: remove extra comma

File: LEAF_Request_Portal/sources/FormEditor.php [hunk 12]
```diff
@@ -581,4 +770,30 @@ private function disableSubindicators($indicatorID)
             $this->setDisabled($item['indicatorID'], 1);
         }
     }
+
+    private function getCategoryID($indicatorID){
+        $vars = array(':indicatorID' => $indicatorID);
+
```
Comment (thomasnguyen95) [inline LEAF_Request_Portal/sources/FormEditor.php]: Please fix spacing

File: LEAF_Request_Portal/utils/db_upgrade/Update_RMC_DB_20200010700-2020011000.sql [hunk 13]
```diff
@@ -0,0 +1,23 @@
+START TRANSACTION;
```
Comment (thomasnguyen95) [inline LEAF_Request_Portal/utils/db_upgrade/Update_RMC_DB_20200010700-2020011000.sql]: please rename file to ```Update_RMC_DB_2020010700-2020011000```
You have 1 extra zero 2020**0**010700

**Comment by dastan-rahimi** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 29
- Added 44
           

Clones removed
==============
+ LEAF_Nexus/sources/Group.php  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/VALEAF/LEAF/pullRequest?prid=4809995&bid=15999835)

File: LEAF_Nexus/db_upgrade/Update_OC_DB_2019121300-2020011000.sql [hunk 14]
```diff
@@ -0,0 +1,18 @@
+START TRANSACTION;
```
[the remark this answers is not in the document]
Comment (g-theriot) [reply LEAF_Nexus/db_upgrade/Update_OC_DB_2019121300-2020011000.sql]: resolved.

File: LEAF_Request_Portal/admin/Group.php (hunk 4 above)
Comment (g-theriot) [reply LEAF_Request_Portal/admin/Group.php]: Updated to match.

File: LEAF_Request_Portal/admin/templates/mod_form.tpl (hunk 6 above)
Comment (g-theriot) [reply LEAF_Request_Portal/admin/templates/mod_form.tpl]: Not really sure.  The only thing I did with these is wrap them so they only make the API call if they've changed.  `visblev` does seem to be an obvious typo.

**Comment by dastan-rahimi** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 29
- Added 33
           

Clones removed
==============
+ LEAF_Nexus/sources/Group.php  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/VALEAF/LEAF/pullRequest?prid=4809995&bid=15999835)

**Comment by dastan-rahimi** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 29
- Added 33
           

Clones removed
==============
+ LEAF_Nexus/sources/Group.php  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/VALEAF/LEAF/pullRequest?prid=4809995&bid=15999835)

**Comment by dastan-rahimi** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 29
- Added 33
           

Clones removed
==============
+ LEAF_Nexus/sources/Group.php  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/VALEAF/LEAF/pullRequest?prid=4809995&bid=15999835)
